stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef STAN_MATH_REV_MAT_FUN_CHOLESKY_DECOMPOSE_HPP
+#define STAN_MATH_REV_MAT_FUN_CHOLESKY_DECOMPOSE_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/mat/fun/cholesky_decompose.hpp>
+#include <stan/math/rev/scal/fun/value_of_rec.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/err/check_pos_definite.hpp>
+#include <stan/math/prim/mat/err/check_square.hpp>
+#include <stan/math/prim/mat/err/check_symmetric.hpp>
+
+namespace stan {
+  namespace math {
+
+    class cholesky_decompose_v_vari : public vari {
+    public:
+      int M_;  // A.rows() = A.cols()
+      vari** _variRefA;
+      vari** _variRefL;
+      vari* _dummy;
+
+      /* ctor for cholesky function
+       *
+       * Stores varis for A
+       * Instantiates and stores varis for L
+       * Instantiates and stores dummy vari for
+       * upper triangular part of var result returned
+       * in cholesky_decompose function call 
+       *
+       * variRefL aren't on the chainable
+       * autodiff stack, only used for storage
+       * and computation. Note that varis for
+       * L are constructed externally in 
+       * cholesky_decompose.
+       *
+       * @param matrix A
+       * @param matrix L, cholesky factor of A
+       * */
+      cholesky_decompose_v_vari(const Eigen::Matrix<var, -1, -1>& A,
+                                const Eigen::Matrix<double, -1, -1>& L_A)
+        : vari(0.0),
+          M_(A.rows()),
+          _variRefA(reinterpret_cast<vari**>
+                    (stan::math::ChainableStack::memalloc_
+                     .alloc(sizeof(vari*) * A.rows() * (A.cols() + 1) / 2))),
+          _variRefL(reinterpret_cast<vari**>
+                    (stan::math::ChainableStack::memalloc_
+                     .alloc(sizeof(vari*) * A.rows() * (A.rows() + 1) / 2))),
+          _dummy(reinterpret_cast<vari*>(stan::math::ChainableStack::memalloc_
```

Comment:
Actually, it probably makes the most sense to just declare a ```var zero(0);``` in the ```cholesky_decompose()``` function and assign it directly instead of allocating it here as part of the vari since it's not needed as part of ```chain()```.